src/super_context.ml

```diff
@@ -33,6 +33,245 @@ module Env_node = struct
+        ; "cxx"            , strings (context.c_compiler :: cxx_flags)
+        ; "ocaml"          , path context.ocaml
+        ; "ocamlc"         , path context.ocamlc
+        ; "ocamlopt"       , path ocamlopt
+        ; "arch_sixtyfour" , string (string_of_bool context.arch_sixtyfour)
+        ; "make"           , make
+        ; "root"           , values [Value.Dir context.build_dir]
+        ] in
+      let uppercased =
+        List.map lowercased ~f:(fun (k, _) ->
+          (String.uppercase k, renamed_in ~new_name:k ~version:(1, 0))) in
+      let vars =
+        [ "-verbose"       , values []
+        ; "pa_cpp"         , strings (context.c_compiler :: cflags
+                                      @ ["-undef"; "-traditional";
+                                         "-x"; "c"; "-E"])
+        ; "ocaml_bin"      , path context.ocaml_bin
+        ; "ocaml_version"  , string context.version_string
+        ; "ocaml_where"    , string (Path.to_string context.stdlib_dir)
+        ; "null"           , string (Path.to_string Config.dev_null)
+        ; "ext_obj"        , string context.ext_obj
+        ; "ext_asm"        , string context.ext_asm
+        ; "ext_lib"        , string context.ext_lib
+        ; "ext_dll"        , string context.ext_dll
+        ; "ext_exe"        , string context.ext_exe
+        ; "profile"        , string context.profile
+        ]
+      in
+      let ocaml_config =
+        List.map (Ocaml_config.to_list context.ocaml_config) ~f:(fun (k, v) ->
+          ("ocaml-config:" ^ k,
+           match (v : Ocaml_config.Value.t) with
+           | Bool   x -> string (string_of_bool x)
+           | Int    x -> string (string_of_int x)
+           | String x -> string x
+           | Words  x -> strings x
+           | Prog_and_args x -> strings (x.prog :: x.args)))
+      in
+      [ ocaml_config
+      ; static_vars
+      ; lowercased
+      ; uppercased
+      ; vars
+      ]
+      |> List.concat
+      |> String.Table.of_list_exn
+
+    let static_vars = String.Table.of_list_exn static_vars
+
+    let rec expand t ~syntax_version ~var =
+      let name =
+        match String_with_vars.Var.destruct var with
+        | Single v -> v
+        | Pair (v, _) -> v
+      in
+      Option.bind (String.Table.find t name) ~f:(function
+        | Nothing v -> Some v
+        | Since (v, min_version) ->
+          if syntax_version >= min_version then
+            Some v
+          else
+            String_with_vars.Var.fail var ~f:(fun var ->
+              sprintf "Variable %s is available in since version %s. \
+                       Current version is %s"
+                var
+                (Syntax.Version.to_string min_version)
+                (Syntax.Version.to_string syntax_version))
+        | Renamed_in (in_version, new_name) -> begin
+            if syntax_version >= in_version then
+              String_with_vars.Var.fail var ~f:(fun old_name ->
```

Comment:
I don't find `Var.fail` particularly clear. I think this code would be clearer if we stick to `Loc.fail (String_with_vars.Var.loc var) "..."`